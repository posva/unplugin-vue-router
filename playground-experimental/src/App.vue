<script setup lang="ts">
import { ref } from 'vue'

const router = useRouter()

console.log(`We have ${router.getRoutes().length} routes.`)

const targetRoute = ref('')
</script>

<template>
  <header>
    <div class="wrapper">
      <nav>
        <ul>
          <li>
            <RouterLink to="/">Home</RouterLink>
          </li>
          <li>
            <RouterLink to="/users/24" #default="{ href }">{{
              href
            }}</RouterLink>
          </li>
          <li>
            <RouterLink to="/events/1992-03-24" #default="{ href }">{{
              href
            }}</RouterLink>
          </li>
        </ul>
      </nav>
    </div>

    <div>
      <p>Currently at "{{ $route.name }}" ({{ $route.fullPath }})</p>
      <pre>{{ $route.params }}</pre>
      <form @submit.prevent="router.push(targetRoute)">
        <label>
          Navigate to:
          <input type="text" v-model="targetRoute" />
        </label>
        <button>Go</button>
      </form>
    </div>
  </header>

  <hr />

  <RouterView />
  <hr />
  <RouterView name="named" />
</template>

<style scoped>
ul {
  padding-left: 0;
}
ul > li {
  display: inline-block;
}

ul > li:not(:first-child) {
  margin-left: 0.5rem;
}

li > a {
  text-decoration: none;
}
.router-link-active {
  text-decoration: underline;
  font-weight: bold;
}
</style>
