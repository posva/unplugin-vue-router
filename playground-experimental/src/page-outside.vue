<script lang="ts" setup></script>

<template>
  <h1>Page outside of <code>pages</code></h1>
</template>
