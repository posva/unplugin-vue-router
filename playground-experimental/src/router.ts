import { experimental_createRouter } from 'vue-router/experimental'
import { resolver, handleHotUpdate } from 'vue-router/auto-resolver'

import {
  type RouteRecordInfo,
  type ParamValue,
  createWebHistory,
} from 'vue-router'

export const router = experimental_createRouter({
  history: createWebHistory(),
  resolver,
})

if (import.meta.hot) {
  handleHotUpdate(router)
}

// manual extension of route types
declare module 'vue-router/auto-routes' {
  export interface RouteNamedMap {
    'custom-dynamic-name': RouteRecordInfo<
      'custom-dynamic-name',
      '/added-during-runtime/[...path]',
      { path: ParamValue<true> },
      { path: ParamValue<false> },
      'custom-dynamic-child-name'
    >
    'custom-dynamic-child-name': RouteRecordInfo<
      'custom-dynamic-child-name',
      '/added-during-runtime/[...path]/child',
      { path: ParamValue<true> },
      { path: ParamValue<false> },
      never
    >
  }
}
