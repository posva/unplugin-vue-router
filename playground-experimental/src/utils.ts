import { inject, toValue, onUnmounted } from 'vue'
import type { RouteLocation, RouteLocationNormalizedLoaded } from 'vue-router'
import { viewDepthKey, useRoute, useRouter } from 'vue-router'
import type { RouteNamedMap } from 'vue-router/auto-routes'

type NavigationReturn = RouteLocation | boolean | void

export function useParamMatcher<Name extends keyof RouteNamedMap>(
  _name: Name,
  fn: (
    to: RouteLocationNormalizedLoaded<Name>
  ) => NavigationReturn | Promise<NavigationReturn>
) {
  const route = useRoute()
  const router = useRouter()
  const depth = inject(viewDepthKey, 0)
  // we only need it the first time
  const matchedRecord = route.matched[toValue(depth) - 1]?.name
  console.log(matchedRecord)

  if (!matchedRecord) return

  console.log('add guard')

  const removeGuard = router.beforeEach((to) => {
    console.log('beforeEach', to)
    if (to.matched.find((record) => record.name === matchedRecord)) {
      return fn(to as RouteLocationNormalizedLoaded<Name>)
    }
  })

  onUnmounted(removeGuard)
}

export function dummy(arg: unknown) {
  return 'ok'
}

export const dummy_id = 'dummy_id'
export const dummy_number = 42
