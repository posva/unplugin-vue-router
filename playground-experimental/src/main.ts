import { createApp } from 'vue'
import App from './App.vue'
import { createPinia } from 'pinia'
import { PiniaColada } from '@pinia/colada'
import { router } from './router'
import { DataLoaderPlugin } from 'unplugin-vue-router/data-loaders'
import { RouterLink, RouterView } from 'vue-router'

const app = createApp(App)

app.use(createPinia())
app.use(PiniaColada, {})
// @ts-expect-error: FIXME: should be doable
app.use(DataLoaderPlugin, { router })
app.component('RouterLink', RouterLink)
app.component('RouterView', RouterView)
app.use(router)

// @ts-expect-error: for debugging on browser
window.$router = router

app.mount('#app')

// small logger for navigations, useful to check HMR
router.isReady().then(() => {
  router.beforeEach((to, from) => {
    console.log('ðŸ§­', from.fullPath, '->', to.fullPath)
  })
})
