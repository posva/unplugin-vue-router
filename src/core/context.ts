import chokidar from 'chokidar'
import { Options } from '../options'
import { createPrefixTree } from './tree'
import { promises as fs } from 'fs'
import { logTree, throttle } from './utils'
import { generateRouteNamedMap } from '../codegen/generateRouteMap'
import { MODULE_ROUTES_PATH, MODULE_VUE_ROUTER } from './moduleConstants'
import { generateRouteRecord } from '../codegen/generateRouteRecords'
import fg from 'fast-glob'
import { resolve } from 'pathe'

export function createRoutesContext(options: Required<Options>) {
  const { dts: preferDTS, root } = options
  const dts =
    preferDTS === false
      ? false
      : preferDTS === true
      ? resolve(root, 'typed-router.d.ts')
      : resolve(root, preferDTS)

  const routeTree = createPrefixTree(options)

  function log(...args: any[]) {
    if (options.logs) {
      console.log(...args)
    }
  }

  const resolvedRoutesFolder = resolve(root, options.routesFolder)
  const serverWatcher = chokidar.watch(resolvedRoutesFolder, {
    ignoreInitial: true,
    disableGlobbing: true,
    ignorePermissionErrors: true,
    // useFsEvents: true,
    // TODO: allow user options
  })

  function stripRouteFolder(path: string) {
    return path.slice(resolvedRoutesFolder.length + 1)
  }

  async function scanPages() {
    const routeFolders: string[] = [resolvedRoutesFolder]
    const pattern = `**/*.{${options.extensions
      .map((extension) => extension.replace('.', ''))
      .join(',')}}`
    const files = (
      await Promise.all(
        routeFolders.map((folder) =>
          fg(pattern, { cwd: folder, followSymbolicLinks: true }).then(
            (files) => files.flatMap((file) => resolve(folder, file))
          )
        )
      )
    ).flat()
    // TODO: filter out excluded files

    for (const file of files) {
      addPage(file)
    }
    await _writeConfigFiles()
  }

  function addPage(path: string) {
    log('added', path)
    const route = stripRouteFolder(path)
    // TODO: handle top level named view HMR
    routeTree.insert(
      route,
      // './' + path
      resolve(root, path)
    )
  }

  function removePage(path: string) {
    log('remove', path)
    routeTree.remove(stripRouteFolder(path))
  }

  function setupWatcher() {
    serverWatcher
      .on('change', (path) => {
        // TODO: parse defineRoute macro?
        log('change', path)
        writeConfigFiles()
      })
      .on('add', (path) => {
        addPage(path)
        writeConfigFiles()
      })
      .on('unlink', (path) => {
        removePage(path)
        writeConfigFiles()
      })
  }

  function generateRoutes() {
    return `export const routes = ${generateRouteRecord(routeTree)}`
  }

  function generateDTS(): string {
    return `
// Generated by unplugin-vue-router. ‼️ DO NOT MODIFY THIS FILE ‼️
// It's recommended to commit this file.
// Make sure to add this file to your tsconfig.json file as an "includes" or "files" entry.

/// <reference types="unplugin-vue-router/client" />

import type {
  _RouterTyped,
  RouteRecordInfo,
  RouterLinkTyped,
  RouteLocationNormalizedLoadedTypedList,
  RouteLocationAsString,
  NavigationGuard,
  _ParamValue,
  _ParamValueOneOrMore,
  _ParamValueZeroOrMore,
  _ParamValueZeroOrOne,
} from 'unplugin-vue-router'

declare module '${MODULE_ROUTES_PATH}' {
${generateRouteNamedMap(routeTree)
  .split('\n')
  .filter((line) => line)
  .map((line) => '  ' + line) // not the same as padStart(2)
  .join('\n')}
}

declare module '${MODULE_VUE_ROUTER}' {
  import type { RouteNamedMap } from '${MODULE_ROUTES_PATH}'

  export type RouterTyped = _RouterTyped<RouteNamedMap>
  /**
   * Generate a type safe route location. Requires the name of the route to be passed as a generic.
   */
  export type Route<Name extends keyof RouteNamedMap> = RouteLocationNormalizedLoadedTypedList<RouteNamedMap>[Name]
  /**
   * Generate a type safe params for a route location. Requires the name of the route to be passed as a generic.
   */
  export type RouteParams<Name extends keyof RouteNamedMap> = RouteNamedMap[Name]['params']
  /**
   * Generate a type safe raw params for a route location. Requires the name of the route to be passed as a generic.
   */
  export type RouteParamsRaw<Name extends keyof RouteNamedMap> = RouteNamedMap[Name]['paramsRaw']

  export function useRouter(): RouterTyped
  export function useRoute<Name extends keyof RouteNamedMap = keyof RouteNamedMap>(name?: Name): RouteLocationNormalizedLoadedTypedList<RouteNamedMap>[Name]

  export function onBeforeRouteLeave(guard: NavigationGuard<RouteNamedMap>): void
  export function onBeforeRouteUpdate(guard: NavigationGuard<RouteNamedMap>): void
}

declare module '@vue/runtime-core' {
  import type { RouteNamedMap } from '${MODULE_ROUTES_PATH}'

  export interface ComponentCustomOptions {
    beforeRouteUpdate?: NavigationGuard<RouteNamedMap>
    beforeRouteLeave?: NavigationGuard<RouteNamedMap>
  }

  export interface ComponentCustomProperties {
    $route: RouteLocationNormalizedLoadedTypedList<RouteNamedMap>[keyof RouteNamedMap]
    $router: _RouterTyped<RouteNamedMap>
  }

  export interface GlobalComponents {
    RouterLink: RouterLinkTyped<RouteNamedMap>
  }
}
`
  }

  // NOTE: this code needs to be generated because otherwise it doesn't go through transforms and `@vue-router/routes`
  // cannot be resolved.
  function generateVueRouterProxy() {
    return `
import { routes } from '@vue-router/routes'
import { createRouter as _createRouter } from 'vue-router'

export * from 'vue-router'

export function createRouter(options) {
  const { extendRoutes } = options
  if (typeof extendRoutes === 'function') {
    routes = extendRoutes(routes) || routes
  }

  return _createRouter({
    ...options,
    routes,
  })
}
`
  }

  let lastDTS: string | undefined
  async function _writeConfigFiles() {
    log('writing')
    logTree(routeTree, log)
    if (dts) {
      const content = generateDTS()
      if (lastDTS !== content) {
        await fs.writeFile(dts, content, 'utf-8')
        lastDTS = content
      }
    }
  }

  // debounce of 100ms + throttle of 500ms
  // => Initially wait 100ms (renames are actually remove and add but we rather write once) (debounce)
  // subsequent calls after the first execution will wait 500ms-100ms to execute (throttling)
  const writeConfigFiles = throttle(_writeConfigFiles, 500, 100)

  setupWatcher()

  function stopWatcher() {
    serverWatcher.close()
  }

  return {
    scanPages,
    writeConfigFiles,

    stopWatcher,

    generateRoutes,
    generateVueRouterProxy,
  }
}
