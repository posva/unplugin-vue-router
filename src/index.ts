import { createUnplugin, type UnpluginOptions } from 'unplugin'
import { createRoutesContext } from './core/context'
import {
  MODULE_ROUTES_PATH,
  MODULE_VUE_ROUTER_AUTO,
  getVirtualId as _getVirtualId,
  asVirtualId as _asVirtualId,
  routeBlockQueryRE,
  ROUTE_BLOCK_ID,
  ROUTES_LAST_LOAD_TIME,
} from './core/moduleConstants'
import {
  Options,
  resolveOptions,
  DEFAULT_OPTIONS,
  mergeAllExtensions,
} from './options'
import { createViteContext } from './core/vite'
import { createFilter } from '@rollup/pluginutils'
import { join } from 'pathe'
import { appendExtensionListToPattern } from './core/utils'
import { MACRO_DEFINE_PAGE_QUERY } from './core/definePage'
import { createAutoExportPlugin } from './data-loaders/auto-exports'

export * from './types'

export { DEFAULT_OPTIONS }

export { AutoExportLoaders } from './data-loaders/auto-exports'
export type { AutoExportLoadersOptions } from './data-loaders/auto-exports'

export default createUnplugin<Options | undefined>((opt = {}, _meta) => {
  const options = resolveOptions(opt)
  const ctx = createRoutesContext(options)

  function getVirtualId(id: string) {
    if (options._inspect) return id
    return _getVirtualId(id)
  }

  function asVirtualId(id: string) {
    // for inspection
    if (options._inspect) return id
    return _asVirtualId(id)
  }

  // create the transform filter to detect `definePage()` inside page component
  const pageFilePattern = appendExtensionListToPattern(
    options.filePatterns,
    mergeAllExtensions(options)
  )

  // this is a larger filter that includes a bit too many files
  // the RouteFolderWatcher will filter it down to the actual files
  const filterPageComponents = createFilter(
    [
      ...options.routesFolder.flatMap((routeOption) =>
        pageFilePattern.map((pattern) => join(routeOption.src, pattern))
      ),
      // importing the definePage block
      /\?.*\bdefinePage\&vue\b/,
    ],
    options.exclude
  )

  const plugins: UnpluginOptions[] = [
    {
      name: 'unplugin-vue-router',
      enforce: 'pre',

      resolveId(id) {
        if (
          // vue-router/auto-routes
          id === MODULE_ROUTES_PATH ||
          // NOTE: it wasn't possible to override or add new exports to vue-router
          // so we need to override it with a different package name
          id === MODULE_VUE_ROUTER_AUTO
        ) {
          // virtual module
          return asVirtualId(id)
        }

        // this allows us to skip the route block module as a whole since we already parse it
        if (routeBlockQueryRE.test(id)) {
          return ROUTE_BLOCK_ID
        }

        // nothing to do, just for TS
        return
      },

      buildStart() {
        return ctx.scanPages(options.watch)
      },

      buildEnd() {
        ctx.stopWatcher()
      },

      // we only need to transform page components
      transformInclude(id) {
        // console.log('filtering ' + id, filterPageComponents(id) ? '‚úÖ' : '‚ùå')
        return filterPageComponents(id)
      },

      transform(code, id) {
        // console.log('üëã  Transforming', id)
        // remove the `definePage()` from the file or isolate it
        return ctx.definePageTransform(code, id)
      },

      // loadInclude is necessary for webpack
      loadInclude(id) {
        if (id === ROUTE_BLOCK_ID) return true
        const resolvedId = getVirtualId(id)
        return (
          resolvedId === MODULE_ROUTES_PATH ||
          resolvedId === MODULE_VUE_ROUTER_AUTO
        )
      },

      load(id) {
        // remove the <route> block as it's parsed by the plugin
        // stub it with an empty module
        if (id === ROUTE_BLOCK_ID) {
          return {
            code: `export default {}`,
            map: null,
          }
        }

        // we need to use a virtual module so that vite resolves the vue-router/auto-routes
        // dependency correctly
        const resolvedId = getVirtualId(id)

        // vue-router/auto-routes
        if (resolvedId === MODULE_ROUTES_PATH) {
          ROUTES_LAST_LOAD_TIME.update()
          return ctx.generateRoutes()
        }

        // vue-router/auto
        if (resolvedId === MODULE_VUE_ROUTER_AUTO) {
          return ctx.generateVueRouterProxy()
        }

        return // ok TS...
      },

      // improves DX
      vite: {
        configureServer(server) {
          ctx.setServerContext(createViteContext(server))
        },

        handleHotUpdate: {
          order: 'post',
          handler({ server, file, modules }) {
            // console.log(`üî• HMR ${file}`)
            const moduleList = server.moduleGraph.getModulesByFile(file)
            const definePageModule = Array.from(moduleList || []).find(
              (mod) => {
                return mod?.id && MACRO_DEFINE_PAGE_QUERY.test(mod.id)
              }
            )

            if (definePageModule) {
              // console.log(`Updating ${definePageModule.file}`)
              const routesModule = server.moduleGraph.getModuleById(
                asVirtualId(MODULE_ROUTES_PATH)
              )

              if (!routesModule) {
                console.error('üî• HMR routes module not found')
                return
              }

              return [
                ...modules,
                // TODO: only if the definePage changed
                definePageModule,
                // TODO: only if ether the definePage or the route block changed
                routesModule,
              ]
            }

            return // for ts
          },
        },
      },
    },
  ]

  // Experimental options
  if (options.experimental.autoExportsDataLoaders) {
    plugins.push(
      createAutoExportPlugin({
        filterPageComponents,
        loadersPathsGlobs: options.experimental.autoExportsDataLoaders,
        root: options.root,
      })
    )
  }

  return plugins
})

export { createRoutesContext }
export { getFileBasedRouteName, getPascalCaseRouteName } from './core/utils'

// Route Tree and edition
export { createTreeNodeValue } from './core/treeNodeValue'
export { EditableTreeNode } from './core/extendRoutes'

/**
 * Adds useful auto imports to the AutoImport config:
 * @example
 * ```js
 * import { VueRouterAutoImports } from 'unplugin-vue-router'
 *
 * AutoImport({
 *   imports: [VueRouterAutoImports],
 * }),
 * ```
 */
export const VueRouterAutoImports: Record<
  string,
  Array<string | [importName: string, alias: string]>
> = {
  'vue-router': [
    'useRoute',
    'useRouter',
    'onBeforeRouteUpdate',
    'onBeforeRouteLeave',
    // NOTE: the typing seems broken locally, so instead we export it directly from unplugin-vue-router/runtime
    // 'definePage',
  ],
  'unplugin-vue-router/runtime': [],
}
