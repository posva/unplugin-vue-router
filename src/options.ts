import { isPackageExists } from 'local-pkg'
import type { TreeLeaf } from './core/tree'
import { getFileBasedRouteName } from './core/utils'

export interface Options {
  extensions?: string[]
  /**
   * Folder containing the components that should be used for routes.
   *
   * @default "src/routes"
   */
  routesFolder?: string

  /**
   * Method to generate the name of a route.
   */
  getRouteName: (node: TreeLeaf) => string

  importMode?:
    | 'sync'
    | 'async'
    | ((path: string, resolvedOptions: Options) => 'sync' | 'async')

  exclude?: string[]

  root?: string

  /**
   * Should generate d.ts files. Defaults to `true` if `typescript` is installed.
   */
  dts?: boolean | string

  /**
   * Allows inspection by vite-plugin-inspect by not adding the leading `\0` to the id of virtual modules.
   * @internal
   */
  _inspect?: boolean

  /**
   * Activates debug logs.
   */
  logs?: boolean
}

export const DEFAULT_OPTIONS: Required<Options> = {
  extensions: ['.vue', '.js', '.jsx', '.ts', '.tsx'],
  exclude: [],
  routesFolder: 'src/routes',
  getRouteName: getFileBasedRouteName,
  importMode: 'async',
  root: process.cwd(),
  dts: isPackageExists('typescript'),
  logs: false,
  _inspect: false,
}
