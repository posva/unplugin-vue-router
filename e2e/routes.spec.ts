import { expect, it } from 'vitest'
import { createRoutesContext } from '../src/core/context'
import { DEFAULT_OPTIONS } from '../src/options'
import { fileURLToPath, URL } from 'url'
import { normalize, join } from 'pathe'

const __dirname = fileURLToPath(new URL('./', import.meta.url))

/**
 * This is a simple full test to check that all filenames are valid in different environment (windows, mac, linux).
 */

it('generates the routes', async () => {
  const context = createRoutesContext({
    ...DEFAULT_OPTIONS,
    // dts: join(__dirname, './__types.d.ts'),
    dts: false,
    logs: false,
    routesFolder: join(__dirname, './fixtures/filenames/routes'),
  })

  await context.scanPages()
  expect(
    context
      .generateRoutes()
      .replace(
        /import\(["'](.+?)["']\)/g,
        (_, filePath) => `import('${normalize(filePath)}')`
      )
      .replace(/(import\(["'])(?:.+?)fixtures\/filenames/gi, '$1')
  ).toMatchSnapshot()
})
