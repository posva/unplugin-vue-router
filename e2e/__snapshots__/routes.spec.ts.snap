// Vitest Snapshot v1

exports[`generates the routes 1`] = `
"export const routes = [
  {
    path: \\"/:path(.*)\\",
    name: \\"/[...path]\\",
    component: () => import('/routes/[...path].vue'),
    /* no children */
  },
  {
    path: \\"/about\\",
    name: \\"/about\\",
    component: () => import('/routes/about.vue'),
    /* no children */
  },
  {
    path: \\"/\\",
    name: \\"/\\",
    component: () => import('/routes/index.vue'),
    /* no children */
  },
  {
    path: \\"/users.new\\",
    name: \\"/users.new\\",
    component: () => import('/routes/users.new.vue'),
    /* no children */
  },
  {
    path: \\"/users\\",
    name: \\"/users\\",
    component: () => import('/routes/users.vue'),
    children: [
      {
        path: \\":id\\",
        name: \\"/users/[id]\\",
        component: () => import('/routes/users/[id].vue'),
        /* no children */
      }
    ],
  },
  {
    path: \\"/articles\\",
    /* no name */
    /* no component */
    children: [
      {
        path: \\":id\\",
        name: \\"/articles/[id]\\",
        component: () => import('/routes/articles/[id].vue'),
        /* no children */
      },
      {
        path: \\":slugs+\\",
        name: \\"/articles/[slugs]+\\",
        component: () => import('/routes/articles/[slugs]+.vue'),
        /* no children */
      }
    ],
  },
  {
    path: \\"/optional\\",
    /* no name */
    /* no component */
    children: [
      {
        path: \\":doc?\\",
        name: \\"/optional/[[doc]]\\",
        component: () => import('/routes/optional/[[doc]].vue'),
        /* no children */
      },
      {
        path: \\":docs*\\",
        name: \\"/optional/[[docs]]+\\",
        component: () => import('/routes/optional/[[docs]]+.vue'),
        /* no children */
      }
    ],
  },
  {
    path: \\"/nested\\",
    /* no name */
    /* no component */
    children: [
      {
        path: \\"folder\\",
        /* no name */
        /* no component */
        children: [
          {
            path: \\"\\",
            name: \\"/nested/folder/\\",
            component: () => import('/routes/nested/folder/index.vue'),
            /* no children */
          },
          {
            path: \\"should\\",
            /* no name */
            /* no component */
            children: [
              {
                path: \\"work\\",
                /* no name */
                /* no component */
                children: [
                  {
                    path: \\"\\",
                    name: \\"/nested/folder/should/work/\\",
                    component: () => import('/routes/nested/folder/should/work/index.vue'),
                    /* no children */
                  }
                ],
              }
            ],
          }
        ],
      }
    ],
  }
]"
`;
